src/electron/helpers/services/comfyui.ts

```diff
@@ -0,0 +1,267 @@
+	private url: string;
+	private host: string;
+	private port: string;
+	private clients: Map<string, WebSocket> = new Map();
+
+	private constructor() {
+		this.host = "127.0.0.1";
+		this.port = "1338";
+		this.url = `${this.host}:${this.port}`;
+	}
+
+	/**
+	 * Initializes the ComfyUI service as a singleton instance.
+	 * This method configures the server, checks its readiness, and starts it if not already running.
+	 *
+	 * @returns {Promise<ComfyUI>} The singleton instance of the ComfyUI class.
+	 * @throws {Error} Throws an error if the server fails to start or is not ready after maximum retries.
+	 */
+	public static async init(): Promise<ComfyUI> {
+		if (!ComfyUI.instance) {
+			ComfyUI.instance = new ComfyUI();
+
+			// Create the config to find the local models of the user
+			await ComfyUI.instance.generateModelPathsConfig();
+
+			// Start the ComfyUI server
+			await ComfyUI.instance.start();
+
+			// Make sure that the ComfyUI server is started
+			const isReady = await ComfyUI.instance.checkServerReady();
+
+			// The server is started
+			if (isReady) {
+				logger.info("ComfyUI is ready");
+			} else {
+				logger.error(`ComfyUI is not ready after maximum retries`);
+			}
+		}
+
+		return ComfyUI.instance;
+	}
+
+	/**
+	 * Generates and writes the model paths configuration file required by the ComfyUI server.
+	 *
+	 * @private
+	 */
+	private async generateModelPathsConfig() {
+		const config = {
+			captain: {
+				base_path: getCaptainDownloads(),
+				checkpoints: "stable-diffusion/checkpoints",
+			},
+		};
+
+		const configYaml = stringify(config);
+
+		createDirectory(getCaptainTemporary("comfyui"));
+		const filePath = getCaptainTemporary("comfyui/model_paths.yaml");
+
+		await fsp.writeFile(filePath, configYaml);
+	}
+
+	/**
+	 * Starts the ComfyUI server using external Python scripts.
+	 * It sets up necessary parameters and handles the process lifecycle.
+	 *
+	 * @private
+	 * @throws {Error} Logs and throws an error if the server process encounters issues.
+	 */
+	private start() {
+		const pythonBinaryPath = getCaptainDownloads("comfyui-portable/python_embeded/python.exe");
+		const scriptPath = getCaptainDownloads("comfyui-portable/ComfyUI/main.py");
+		const modelPaths = getCaptainTemporary("comfyui/model_paths.yaml");
+		const outputPath = getCaptainData("files/comfyui");
+
+		createDirectory(outputPath);
+
+		const scriptArguments = [
+			"--listen",
+			this.host,
+			"--port",
+			this.port,
+			"--disable-auto-launch",
+			"--extra-model-paths-config",
+			modelPaths,
+			"--output-directory",
+			outputPath,
+		];
+
+		// Start ComfyUI process
+		this.process = execa(pythonBinaryPath, ["-u", scriptPath, ...scriptArguments]);
+
+		// Listen for stdout and stderr
+		if (this.process.stdout && this.process.stderr) {
+			this.process.stdout.on("data", async data => {
+				logger.info(data.toString());
+			});
+
+			this.process.stderr.on("data", data => {
+				logger.info(data.toString());
```

Comment:
which type can data be? make sure we don't try to cast strings from objects.

Reply:
I added a type.